libraries/libfc/include/fc/network/listener.hpp

```diff
@@ -32,94 +32,50 @@ inline std::pair<std::string, std::string> split_host_port(std::string_view endp
-///    using socket_type = Protocol::socket;
-///    socket_type&& socket_;
-///    shared_state_type& shared_state_;
-///    example_session(socket_type&& socket, shared_state_type& shared_state)
-///      : socket_(std::move(socket)), shared_state_(shared_state_) {}
-///
-///    // ...
-///    void start();
-/// };
-///
-/// template <typename Protocol>
-/// struct example_listener : fc::listener<example_listener<Protocol>, Protocol>{
-///    static constexpr uint32_t accept_timeout_ms = 200;
-///    shared_state_type& shared_state_;
-///
-///    example_listener(boost::asio::io_context& executor,
-///                     logger& logger,
-///                     const std::string& local_address,
-///                     const typename Protocol::endpoint& endpoint,
-///                     shared_state_type& shared_state)
-///    : fc::listener<example_listener<Protocol>, Protocol>
-///        (executor, logger, boost::posix_time::milliseconds(accept_timeout_ms), local_address, endpoint)
-///    , shared_state_(shared_state) {}
-///
-///    std::string extra_listening_log_info() {
-///       return shared_state_.info_to_be_printed_after_address_is_resolved_and_listening;
-///    }
-///
-///    void create_session(Protocol::socket&& sock) {
-///        auto session = std::make_shared<example_session>(std::move(sock), shared_state_);
-///        session->start();
-///    }
-///  };
-///
-///  int main() {
-///    boost::asio::io_context ioc;
-///    fc::logger logger = fc::logger::get(DEFAULT_LOGGER);
-///    shared_state_type shared_state{...};
-///
-///    // usage for accepting tcp connection
-///    // notice that it only throws std::system_error, not fc::exception
-///    example_listener<boost::asio::ip::tcp>::create(executor, logger, "localhost:8080", std::ref(shared_state));
-///
-///    // usage for accepting unix socket connection
-///    example_listener<boost::asio::local::stream_protocol>::create(executor, logger, "tmp.sock",
-///    std::ref(shared_state));
-///
-///    ioc.run();
-///    return 0;
-///  }
-/// \endcode
+/// @note Users should use fc::create_listener() instead, this class is the implementation
+/// detail for fc::create_listener().
 ///
 /////////////////////////////////////////////////////////////////////////////////////////////
-
-template <typename T, typename Protocol>
-struct listener : std::enable_shared_from_this<T> {
+template <typename Protocol, typename CreateSession>
+struct listener : listener_base<Protocol>, std::enable_shared_from_this<listener<Protocol, CreateSession>> {
    using endpoint_type = typename Protocol::endpoint;
 
    typename Protocol::acceptor      acceptor_;
    boost::asio::deadline_timer      accept_error_timer_;
    boost::posix_time::time_duration accept_timeout_;
    logger&                          logger_;
-   std::string                      local_address_;
+   std::string                      extra_listening_log_info_;
+   CreateSession                    create_session_;
```

Comment:
You can make those members that are *not* accessed outside the class private, so you wouldn't need to add getter/setter functions. This would comply better with [this](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-private) C++ guideline.

Reply:
Can you make those members that are not accessed outside the class private?

Reply:
Done.